src/coprocessor/dag/batch/executors/util/aggr_executor.rs

```diff
@@ -68,11 +70,22 @@ pub trait AggregationExecutorImpl<Src: BatchExecutor>: Send {
     ///
     /// Implementors may return the content of each group as extra columns in the return value
     /// if there are group by columns.
-    fn iterate_each_group_for_aggregation(
+    ///
+    /// Implementors should not iterate the same group multiple times for the same partial
+    /// input data.
+    fn iterate_each_group_for_partial_aggregation(
         &mut self,
         entities: &mut Entities<Src>,
+        src_is_drained: bool,
         iteratee: impl FnMut(&mut Entities<Src>, &[Box<dyn AggrFunctionState>]) -> Result<()>,
     ) -> Result<Vec<LazyBatchColumn>>;
+
+    /// Returns whether we can aggregate at the moment.
+    ///
+    /// The default value is `src_is_drained`. Only StreamAgg can aggregate when not drained.
+    fn is_partial_results_ready(&self, src_is_drained: bool) -> bool {
```

Comment:
For stream agg, hash agg, simple agg, all of them, and maybe other kind of aggregation in future, when `src_is_drained`, partial results are always ready. Thus maybe we can simply extract the `src_is_drained` parameter from this function signature. Also, after the extraction looks like it would be better to leave this function *without* default impl, since this executor should have no knowledge about the implementor. We know that only one executor (stream agg) supports partial results, but this executor should not keep that knowledge.